server/src/main/java/org/elasticsearch/index/mapper/FieldTypeLookup.java

```diff
@@ -79,9 +79,15 @@
         }
 
         for (RuntimeField runtimeField : runtimeFields) {
+            if (runtimeField instanceof DynamicFieldType) {
+                dynamicFieldTypes.put(runtimeField.name(), (DynamicFieldType) runtimeField);
+            }
             MappedFieldType runtimeFieldType = runtimeField.asMappedFieldType();
-            //this will override concrete fields with runtime fields that have the same name
-            fullNameToFieldType.put(runtimeFieldType.name(), runtimeFieldType);
+            assert runtimeFieldType != null || runtimeField instanceof DynamicFieldType;
```

Comment:
Do we expect that dynamic runtime fields will always return `null` here? In which case maybe this should just be in a separate branch of the `if (runtimeField instanceof DynamicFieldType)` test above?

Reply:
but the assert also checks that when it's not dynamic it is never null, just to make sure. I am not extremely sure what the expectations are. We could also not call asMappedFieldType for dynamic field types, and be done with it, then what it returns does not really matter. That is subtle but we could make it throw UEO.